### ISSUE TITLE
Bulk entities retrieval returns confusing properties of each entity

### ISSUE THREAD

**Issue by gegnew** [primitybio/cellengine-r-toolkit#48]
When you retrieve entities  in bulk (i.e. `getFcsFiles`), and then get a single entity from the list (see code below), then a property that is a list (like annotations) are returned as a nested list.

```
files = getFcsFiles(id)
f = files[1,]
f$annotations
# prints:
[[1]]
           name value
1 annotations 1     1
2  annotation 2 value
```

However,

```
f = getFcsFile(id, byName("myGreatFcsFile")
f$annotations
# prints:
           name value
1 annotations 1     1
2  annotation 2 value
```

**Comment by gegnew** [comment]
This is so confusing. `getFcsFile` returns a `list`, but `getFcsFiles` returns a `data.frame`. However, `getFcsFile()$annotations` returns a `data.frame`, and `getFcsFiles()$annotations` returns the annoying nested list:

```r
> f = getFcsFiles()[1,]
> f2 = getFcsFile()
> class(f$annotations)
[1] "list"
> class(f2$annotations)
[1] "data.frame"
> class(f)
[1] "data.frame"
> class(f2)
[1] "list"
```

**Comment by zbjornson** [comment]
It's worth making a breaking change if it means we can provide a nicer, consistent experience.

On that note, we need to start keeping a changelog and following semver. I like to use https://keepachangelog.com/en/1.0.0/ this format -- notably, keeping an `## Unreleased` section at the top with areas to log breaking changes, additions and fixes.

**Comment by gegnew** [comment]
It's possible that [tidyjson](https://mran.microsoft.com/snapshot/2017-02-04/web/packages/tidyjson/vignettes/introduction-to-tidyjson.html) would be a better option; I'll look into it after I finish the flowCore conversions

**Comment by zbjornson** [comment]
Our ScaleSets are awkward also, for some reason `scales` is a length-1 list:

```R
scaleSet = getScaleSets(experimentId)
chanIdx = 5
applyScale(scaleSet$scales[[1]][chanIdx, "scale"], c(1, 2, 3, 4, 5))
```

**Comment by zbjornson** [comment]
A response was posted to that issue that looks like a valid solution.

**Comment by gegnew** [comment]
To summarize the issue:
- `jsonlite` returns a `list` for single entitles (`getFcsFile`)
- it returns a `data.frame` for bulk entities
- indexing for a single row of that `data.frame` (`files[1,]`) returns a `data.frame`, not a list
- `as.list(files[1,])` and similar methods returning objects which themselves contain data.frames have the data.frame inside a list (`files[1,]$panel == [['a data.frame']]`, but `getFcsFile('id')$panel == 'a data.frame')
- `unlist(files[1,], recursive=F)` returns an equivalent list to single entity retrieval, without the nested list
- There does not appear to be any way to force the "nice" format (a `list` that simply contains dataframes where necessary) onto a bulk-retrieval data.frame
- I _could_ unpack the bulk-retrieval data.frame, `unlist` each entity, and then return the entities in a list, but this eliminates the usefulness of getting a data.frame in the first place


In short, I don't really think there's much to be done. I'd add the `unlist` method to the docs, but otherwise I'm not sure what the path forward is

**Comment by zbjornson** [comment]
Sorry for a late reply.

I keep finding more issues with the current decoding. The latest one is the (in retrospect, obvious) realization that all elements in a dataframe have to have the same properties. Example issue:
1. Make two experiments.
2. In one of them, add an annotation from the header (pick any). Server-side, this adds a property to the `annotationSources` object.
3. Call `getExperiments()`. *All* experiments in the dataframe now have an `annotationSources.XYZ` property now.

I think `getFcsFiles()` and other "list" functions should be returning a list of lists, and eventually a list of S3 objects. It'll be harder to extract a list/vector of properties from all objects (like a list of filenames), but I'm convinced dataframes are the wrong object type.

If you agree, we can close #84 since that's the opposite direction.

**Comment by zbjornson** [comment]
As of the above commit, all methods return lists of lists. Nested dataframes become top-level properties (e.g. `annotations.row` from `getStatistics()`).
